src/utils.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+from functools import wraps
+from typing import Any, Callable, Optional
+
+from constants import GLAUTH_CONFIG_FILE
+from ops.charm import CharmBase, EventBase
+from ops.model import BlockedStatus, WaitingStatus
+from tenacity import Retrying, TryAgain, wait_fixed
+
+logger = logging.getLogger(__name__)
+
+
+def leader_unit(func: Callable) -> Callable:
+    @wraps(func)
+    def wrapper(charm: CharmBase, *args: Any, **kwargs: Any) -> Optional[Any]:
+        if not charm.unit.is_leader():
+            return None
+
+        return func(charm, *args, **kwargs)
+
+    return wrapper
+
+
+def validate_container_connectivity(func: Callable) -> Callable:
+    @wraps(func)
+    def wrapper(charm: CharmBase, *args: EventBase, **kwargs: Any) -> Optional[Any]:
+        event, *_ = args
+        logger.debug(f"Handling event: {event}")
+        if not charm._container.can_connect():
+            logger.debug(f"Cannot connect to container, defer event {event}.")
+            event.defer()
+
+            charm.unit.status = WaitingStatus("Waiting to connect to container.")
+            return None
+
+        return func(charm, *args, **kwargs)
+
+    return wrapper
+
+
+def validate_integration_exists(integration_name: str) -> Callable:
+    def decorator(func: Callable) -> Callable:
+        @wraps(func)
+        def wrapper(charm: CharmBase, *args: EventBase, **kwargs: Any) -> Optional[Any]:
+            event, *_ = args
+            logger.debug(f"Handling event: {event}")
+
+            if not charm.model.relations[integration_name]:
+                logger.debug(f"Integration {integration_name} is missing, defer event {event}.")
+                event.defer()
+
+                charm.unit.status = BlockedStatus(
+                    f"Missing required integration {integration_name}"
+                )
+                return None
+
+            return func(charm, *args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+def validate_database_resource(func: Callable) -> Callable:
+    @wraps(func)
+    def wrapper(charm: CharmBase, *args: EventBase, **kwargs: Any) -> Optional[Any]:
+        event, *_ = args
+        logger.info(f"Handling event: {event}")
+
+        if not charm.database_requirer.is_resource_created():
+            logger.info(f"Database has not been created yet, defer event {event}")
+            event.defer()
+
+            charm.unit.status = WaitingStatus("Waiting for database creation")
+            return None
+
+        return func(charm, *args, **kwargs)
+
+    return wrapper
+
+
+def after_config_updated(func: Callable) -> Callable:
```

Comment:
I use this function to resolve the configmap propagation delay issue in k8s for now since the config file (as a piece of data, will eventually reach a steady state).

I was thinking of adding a timeout to the loop but did not come up with a number. Let me know if you have any thoughts or suggestions.

Reply:
Perhaps it would make sense to set the charm to a waiting state, so the user knows what is happening.

Not sure if this makes sense 100%, because the app will be functional as the previous version will be running. Still I think that we need to tell the user that the delay is normal and expected, because they might think that there is something wrong with the deployment if an action takes  ~40secs to run. Of course this is not be necessary in all cases.